crates/protocol/src/batch/traits.rs

```diff
@@ -23,3 +26,24 @@ pub trait BatchValidationProvider {
     /// Errors if no block is available for the given block number.
     async fn block_by_number(&mut self, number: u64) -> Result<OpBlock, Self::Error>;
 }
+
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+impl<N, T, P> BatchValidationProvider<N, T> for P
+where
+    N: Network,
+    T: Transport + Clone,
+    P: Provider<T, N>,
+{
+    type Error = TransportError;
+
+    async fn l2_block_info_by_number(&mut self, number: u64) -> Result<L2BlockInfo, Self::Error> {
+        // TODO: Actual request + logic
+        self.client().request("eth_getBlockByNumber", (number,)).await
+    }
+
+    async fn block_by_number(&mut self, number: u64) -> Result<OpBlock, Self::Error> {
+        // TODO: Actual request + logic
+        self.client().request("eth_getBlockByNumber", (number, true)).await
+    }
+}
```

Comment:
The reason why the `BatchValidationProvider` trait is defined here is to abstract away the provider logic from the protocol crate. That allows for any downstream consumers of the `maili-protocol` crate to define their own way of providing this data fetching. For example, [`BatchValidationProvider` is implemented](https://github.com/op-rs/kona/blob/3f001c95c79ef0c44eae6dd28ab3ceb0e992a6f1/crates/proof-sdk/proof/src/l2/chain_provider.rs#L75) for the provider that is used when validating batches. Let me know if this makes sense!

Reply:
Thanks @refcell that makes a lot of sense. I see now the issue description was clear about implementing the trait for `ReqwestProvider`. Have pushed a small change to that effect `impl BatchValidationProvider for alloy_provider::ReqwestProvider {`.

Will proceed with the rest of the implementation later today.

Reply:
Got it thanks for clarifying. I'm supportive of these changes, but can we move this implementation into a `reqwest.rs` file, and hide the module behind a `reqwest` feature flag?

e.g. in `lib.rs` we could do the following.

```rust
...

#[cfg(feature = "reqwest")]
pub mod reqwest;

...
```